src/dtx/dtx_common.c

```diff
@@ -296,35 +296,29 @@ dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
 	return 0;
 }
 
+static int
+dtx_wait(struct dtx_handle *dth, struct dtx_conflict_entry **dces);
+
 int
 dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
 	struct ds_cont_child *cont, int result)
 {
-	int	rc = 0;
-
-	if (result < 0) {
-		if (!dth->dth_leader && dth->dth_dti_cos_count > 0) {
-			/* XXX: For non-leader replica, even if we fail to
-			 *	make related modification for some reason,
-			 *	we still need to commit the DTXs for CoS.
-			 *	Because other replica may have already
-			 *	committed them. For leader case, it is
-			 *	not important even if we miss to commit
-			 *	the CoS DTXs, because they are still in
-			 *	CoS cache, and can be committed next time.
-			 */
-			rc = vos_dtx_commit(cont->sc_hdl, dth->dth_dti_cos,
-					    dth->dth_dti_cos_count);
-			if (rc != 0)
-				D_ERROR(DF_UUID": Fail to DTX CoS commit: %d\n",
-					DP_UUID(cont->sc_uuid), rc);
-		}
+	struct dtx_conflict_entry	*dces = NULL;
+	int				rc = 0;
 
-		goto fail;
-	}
+	if (dth == NULL)
+		return result;
 
-	if (!dth->dth_leader || dth->dth_non_rep ||  dtx_is_null(dth->dth_ent))
-		goto out;
+	if (!dth->dth_leader || dth->dth_non_rep ||  dtx_is_null(dth->dth_ent) ||
+	    dth->dth_exec_arg == NULL)
+		goto out_free;
+
+	rc = dtx_wait(dth, &dces);
```

Comment:
did not see where to use/free the returned dces? are you plan to do the dtx_conflict() in other following patch?

Reply:
ah, right. sigh. these seems some conflict. :(